lib/mercury/cps.rb

```diff
@@ -0,0 +1,97 @@
+require 'mercury/cps/seq'
+require 'mercury/cps/seq_with_let'
+require 'mercury/cps/methods'
+require 'mercury/utils'
+
+# Async IO often involves CPS (continuation-passing style)
+# code, where the continuation (a.k.a. "callback") is passed
+# to a function to be invoked at a later time. CPS style
+# can result in deep lexical nesting making code difficult
+# to read.
+# This monad hides the CPS plumbing, which allows code
+# to be written in a flat style with no visible continuation
+# passing. It basically wraps a CPS proc with methods for
+# composing them.
+# See http://codon.com/refactoring-ruby-with-monads
+class Cps
+  attr_reader :cps
+
+  # @param [Proc] cps a CPS proc (signature: *args, &k)
+  def initialize(&cps)
+    @cps = cps
+  end
+
+  # Applies the wrapped proc. If the CPS return value
+  # is not needed, the continuation k may be omitted.
+  # Returns the return value of the continuation.
+  def run(*args, &k)
+    k ||= proc{|x| x}
+    cps.call(*args, &k)
+  end
+
+  # The "bind" operation; composes two Cps
+  # @param [Proc] pm a proc that takes the output of this
+  #   Cps and returns a Cps
+  def and_then(&pm)
+    Cps.new do |*args, &k|
+      self.run(*args) do |*args2|
+        next_cps = pm.call(*args2)
+        next_cps.is_a?(Cps) or raise "`and_then` block did not return a Cps. Did you want `and_lift`? at #{pm.source_location}"
```

Comment:
2 recommendations.

1. due to length of line, put the message into a variable
2. I would recommend using single quotes `'` instead of back quotes `\`` in the message itself